File: 240430/둘 중 하나 잘 고르기/choose-one-of-two-points.py
n = int(input())
dp = [[[0 for _ in range(n+1)] for _ in range(n+1)] for _ in range(2*n+1)]
for i in range(1, 2*n+1):
    red, blue = map(int, input().split())
    for j in range(n+1):
        for k in range(n+1):
            if j > 0:
                dp[i][j][k] = max(dp[i][j][k], dp[i-1][j-1][k] + red)
            if k > 0:
                dp[i][j][k] = max(dp[i][j][k], dp[i-1][j][k-1] + blue)
            dp[i][j][k] = max(dp[i][j][k], dp[i-1][j][k])
print(dp[2*n][n][n])

File: 260830/둘 중 하나 잘 고르기/choose-one-of-two-points.java
import java.util.*;
public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] red = new int[2*n+1];
        int[] blue = new int[2*n+1];
        for (int i = 1; i <= 2 * n; i++) {
            red[i] = sc.nextInt();
            blue[i] = sc.nextInt();
        }
        int[][] dp = new int[n+1][n+1];
        for(int i = 0; i <= n; i++)
            Arrays.fill(dp[i], -1);
        dp[0][0] = 0;
        for(int i = 1; i <= n; i++){
            dp[i][0] = Math.max(dp[i][0], dp[i-1][0] + red[i]);
            dp[0][i] = Math.max(dp[0][i], dp[0][i-1] + blue[i]);
        }
        for(int i = 1; i <= n; i++){
            for(int j = 1; j <= n; j++){
                if(dp[i-1][j] != -1)
                    dp[i][j] = Math.max(dp[i][j], dp[i-1][j] + red[i+j]);
                if(dp[i][j-1] != -1)
                    dp[i][j] = Math.max(dp[i][j], dp[i][j-1] + blue[i+j]);
            }
        }
        System.out.println(dp[n][n]);
    }
}